Modified AutomotiveSimulator To Display Prius Models Instead of Boxes

liangfok opened the pull request:
Builds upon #3921. This goes towards spiral 1 of #3919.

It replaces the "box car" with a model of the Prius:

<img width="1126" alt="screen shot 2016-10-29 at 3 47 42 pm" src="https://cloud.githubusercontent.com/assets/1388098/19832324/41a7acfe-9def-11e6-8a16-261e5606f115.png">

<!-- Reviewable:start -->
---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/3933)

<!-- Reviewable:end -->

liangfok commented:
+@maddog-tri for feature review please.
+@jwnimmer-tri for platform review please.

Thanks!

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-:-KV6RnyTVVwz1m6kuzUr:b-kigsmr)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
-@jwnimmer-tri until #3921 is merged.  Don't want to re-review changes.  Feel free to tag me once 3921 is in and this branch has been updated.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-:-KV6_TmS3rb-tR0otmyJ:bld4ya3)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Will close until spiral 1 of #3919 is done.

liangfok commented:
This PR passed all CI checks. Thus +@jwnimmer-tri for platform review please. 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-:-KVHIxOIrU_VPKYVJESb:bg3sfvl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/automotive/automotive_simulator.cc, line 246 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVHJqFPCpC2M6SAlqxj:-KVHJqFQ8FcPxVYv2LvZ:b91zqbk) ([raw file](https://github.com/robotlocomotion/drake/blob/19aaebf9fcafaeba30eeb2b6f2eedc6cd4a549bd/drake/automotive/automotive_simulator.cc#L246)):*
> ```cpp
> 
>     // Connects systems that provide joint positions to the mux position inputs.
>     // Connect the zero-velocity source to all of the mux velocity inputs.
> ```

Connects.

---

*[drake/automotive/automotive_simulator.cc, line 254 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVHKE5GwuXEC424ZAwj:-KVHKE5GwuXEC424ZAwk:b32whe) ([raw file](https://github.com/robotlocomotion/drake/blob/19aaebf9fcafaeba30eeb2b6f2eedc6cd4a549bd/drake/automotive/automotive_simulator.cc#L254)):*
> ```cpp
> 
>       // Verifies that the current model instance is connected to the world via
>       // a RPY floating joint.
> ```

an

---

*[drake/automotive/automotive_simulator.cc, line 270 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVHKYqZ3W4k34JmOcZx:-KVHKYq_zo7qmxwMSvpY:bb64e1a) ([raw file](https://github.com/robotlocomotion/drake/blob/19aaebf9fcafaeba30eeb2b6f2eedc6cd4a549bd/drake/automotive/automotive_simulator.cc#L270)):*
> ```cpp
>       DRAKE_DEMAND(base_joint.get_num_velocities() == kRpyJointNumVel);
> 
>       // Determines the number of DOFs in the model
> ```

Missing period.

---

*[drake/automotive/automotive_simulator.h, line 47 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVHPEOyb4RT3cEJy0d9:-KVHPEOzBicTeyrIRK0r:b6xuy4) ([raw file](https://github.com/robotlocomotion/drake/blob/19aaebf9fcafaeba30eeb2b6f2eedc6cd4a549bd/drake/automotive/automotive_simulator.h#L47)):*
> ```objc
>   const RigidBodyTree& get_rigid_body_tree();
> 
>   /// Adds a box-shaped SimpleCar system to this simulation, including its
> ```

No longer box-shaped.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3933)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/automotive/automotive_simulator.cc, line 246 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVHJqFPCpC2M6SAlqxj:-KVHio6qFRV0x_k-48qb:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/19aaebf9fcafaeba30eeb2b6f2eedc6cd4a549bd/drake/automotive/automotive_simulator.cc#L246)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Connects.
</details>
Done.

---

*[drake/automotive/automotive_simulator.cc, line 254 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVHKE5GwuXEC424ZAwj:-KVHiqlafsB4EokbcnuS:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/19aaebf9fcafaeba30eeb2b6f2eedc6cd4a549bd/drake/automotive/automotive_simulator.cc#L254)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> an
</details>
Done.

---

*[drake/automotive/automotive_simulator.cc, line 270 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVHKYqZ3W4k34JmOcZx:-KVHiuCdTX4qwaofW2_D:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/19aaebf9fcafaeba30eeb2b6f2eedc6cd4a549bd/drake/automotive/automotive_simulator.cc#L270)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Missing period.
</details>
Done.

---

*[drake/automotive/automotive_simulator.h, line 47 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVHPEOyb4RT3cEJy0d9:-KVHj06GgCwqcnTmABRY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/19aaebf9fcafaeba30eeb2b6f2eedc6cd4a549bd/drake/automotive/automotive_simulator.h#L47)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> No longer box-shaped.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3933)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 2 files at r3, 4 of 5 files at r4.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[drake/automotive/automotive_simulator.cc, line 207 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVK_3OmuwmZYIf5VG6a:-KVK_3Onu8Uutf4lE0fE:bcpfrln) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/automotive_simulator.cc#L207)):*
> ```cpp
>     int num_joint_positions{0};
>     int num_joint_velocities{0};
>     const std::vector<const RigidBody*> bodies
> ```

BTW I think its more conventional to put the `=` on the prior line.  Does clang-format not care?

---

*[drake/automotive/automotive_simulator.cc, line 216 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKZtfnjsgWdwoCKVbY:-KVKZtfnjsgWdwoCKVbZ:bhuauu) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/automotive_simulator.cc#L216)):*
> ```cpp
>     velocity_state_vec.push_back(num_joint_velocities);
>   }
>   std::vector<int> result;
> ```

BTW Something like this would be more conventional:
http://stackoverflow.com/questions/3177241/what-is-the-best-way-to-concatenate-two-vectors

---

*[drake/automotive/automotive_simulator.cc, line 278 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKb0fok3YYU1f0ZJUf:-KVKb0fpEmtnpF-Rt-zy:b-lzpotu) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/automotive_simulator.cc#L278)):*
> ```cpp
>         num_position_dofs += body->getJoint().get_num_positions();
>         num_velocity_dofs += body->getJoint().get_num_velocities();
>       }
> ```

Shouldn't this block reuse the `GetModelJointStateSizes` result, instead of independently reimplementing it?

---

*[drake/automotive/automotive_simulator.cc, line 302 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKbZ1JGzIo-qFmV06e:-KVKbZ1JGzIo-qFmV06f:brwnbix) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/automotive_simulator.cc#L302)):*
> ```cpp
>         auto position_mux =
>             builder_->template AddSystem<systems::Multiplexer<T>>(
>                 position_mux_port_sizes);
> ```

BTW Something like this would be less verbose (using vector's initializer_list):
```
auto position_mux = builder_->template AddSystem<systems::Multiplexer<T>>(
    std::vector<int>{kRpyJointNumPos, num_reg_pos_dofs});
```

---

*[drake/automotive/automotive_simulator.cc, line 309 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKbLSJP8rDwRNszPlT:-KVKbLSJP8rDwRNszPlU:b-5ko8uq) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/automotive_simulator.cc#L309)):*
> ```cpp
>         builder_->Connect(position_mux->get_output_port(0),
>           multiplexer->get_input_port(model_index));
>       }
> ```

FYI The _GeometryWorld_'s appearance cannot come fast enough for me, to remove the need for ghastly logic like this stuff.  Wow!

---

*[drake/automotive/automotive_simulator.cc, line 319 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVK_lA56v_tMtLpeHyC:-KVK_lA6x1XsZOrvgn1M:b-x0khu1) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/automotive_simulator.cc#L319)):*
> ```cpp
>           multiplexer->get_input_port(num_models + model_index));
>     }
>   }
> ```

Probably even the version on `master` previously deserved a helper method for this block, to improve readability of Start().  This new rewrite definitely does.

---

*[drake/automotive/automotive_simulator.cc, line 320 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVK_xZ__lmEvxDE4x81:-KVK_xZ__lmEvxDE4x82:bm8q9if) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/automotive_simulator.cc#L320)):*
> ```cpp
>     }
>   }
>   rigid_body_tree_->compile();
> ```

BTW Are we sure that calling `compile()` _after_ all of the bare-integer math above is sound?  If so, please clarify why in a comment.  If not, consider adding a TODO or assertion highlighting the hazard?

---

*[drake/automotive/automotive_simulator.h, line 111 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKYCBsPsfHlNb8u0R-:-KVKYCBsPsfHlNb8u0R0:bviuwhz) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/automotive_simulator.h#L111)):*
> ```objc
>  private:
>   int allocate_vehicle_number();
>   void AddSdfModel(const std::string sdf_filename,
> ```

Don't use `const` on pass-by-value.

---

*[drake/automotive/automotive_simulator.h, line 120 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKYN3v177KD47oMqbJ:-KVKYN3wIeatUbl8rvFd:b5wujhw) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/automotive_simulator.h#L120)):*
> ```objc
>   // model instances since each model instance has two entries: (1) its number
>   // of position states and (2) its number of velocity states.
>   std::vector<int> GetModelJointStateSizes();
> ```

BTW The API comment reads like a `const` method, but its not marked const.

---

*[drake/automotive/automotive_simulator.h, line 130 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKYj1Cm5zKs7ODwPze:-KVKYj1DZmkjWrOi3iRG:b-vmvkfv) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/automotive_simulator.h#L130)):*
> ```objc
>   std::unique_ptr<systems::DiagramBuilder<T>> builder_{
>     std::make_unique<systems::DiagramBuilder<T>>()};
> ```

BTW Given this new blank line and the one following, consider if there's a way to restore the clarity that the "For building" comment applies to a group of several member fields, e.g. `// === For building ===` or something.

---

*[drake/automotive/test/automotive_simulator_test.cc, line 59 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKbz0-POvwwCGl1ECv:-KVKbz00f9y3wQbhxQqV:b-3795qi) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/test/automotive_simulator_test.cc#L59)):*
> ```cpp
>   const auto& tree = simulator->get_rigid_body_tree();
>   EXPECT_EQ(1, tree.get_num_model_instances());
>   ASSERT_EQ(18, tree.get_num_bodies());  // (0 is world, 17 is part of a Prius.)
> ```

BTW "17 are" typo, I think?

---

*[drake/automotive/test/automotive_simulator_test.cc, line 61 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKcIUyfwfuzlsw7qoO:-KVKcIUyfwfuzlsw7qoP:b-ycrr3b) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/test/automotive_simulator_test.cc#L61)):*
> ```cpp
>   ASSERT_EQ(18, tree.get_num_bodies());  // (0 is world, 17 is part of a Prius.)
>   const auto& body = tree.get_body(1);
>   EXPECT_EQ("chassis_floor", body.get_name());
> ```

FYI Once the API allows selecting the SDF / URDF file again, I think the unit test was superior with a simpler model under test in the base cases (and then add new case for a multibody model).  Consider a TODO for fixing that in the next iteration?

---

*[drake/automotive/test/automotive_simulator_test.cc, line 132 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKcCS-_cxraYQ3CISb:-KVKcCS-_cxraYQ3CISc:b-apc8aw) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/test/automotive_simulator_test.cc#L132)):*
> ```cpp
>   const auto& tree = simulator->get_rigid_body_tree();
>   EXPECT_EQ(2, tree.get_num_model_instances());
>   ASSERT_EQ(35, tree.get_num_bodies());  // (0 is world, 1 & 2 are Prius cars.)
> ```

Rule-of-3 on missing name for magic literal "17". 

---

*[drake/automotive/test/automotive_simulator_test.cc, line 134 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKc6T_IBTlYP2_krk6:-KVKc6T_IBTlYP2_krk7:bqfg3r0) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/test/automotive_simulator_test.cc#L134)):*
> ```cpp
>   ASSERT_EQ(35, tree.get_num_bodies());  // (0 is world, 1 & 2 are Prius cars.)
>   EXPECT_EQ("chassis_floor", tree.get_body(1).get_name());
>   EXPECT_EQ("front_axle", tree.get_body(2).get_name());
> ```

Please amend this line so that it continues to test that the second car got added successfully, not the second body on the first car.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3933)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@jwnimmer-tri: Thanks for the review. I addressed your comments. PTAL.

---

Review status: 2 of 5 files reviewed at latest revision, 14 unresolved discussions.

---

*[drake/automotive/automotive_simulator.cc, line 207 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVK_3OmuwmZYIf5VG6a:-KVKmEDPwLyP2WBBGyMA:b-dr2egi) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/automotive_simulator.cc#L207)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW I think its more conventional to put the `=` on the prior line.  Does clang-format not care?
</details>
Done. I ran `clang-format` on this file and on its header file.

---

*[drake/automotive/automotive_simulator.cc, line 216 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKZtfnjsgWdwoCKVbY:-KVKmtQ54MKctFc48RTU:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/automotive_simulator.cc#L216)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Something like this would be more conventional:
> http://stackoverflow.com/questions/3177241/what-is-the-best-way-to-concatenate-two-vectors
</details>
Done.

---

*[drake/automotive/automotive_simulator.cc, line 278 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKb0fok3YYU1f0ZJUf:-KVKnsvoU9lryWp5qIpq:b3krj) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/automotive_simulator.cc#L278)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Shouldn't this block reuse the `GetModelJointStateSizes` result, instead of independently reimplementing it?
</details>
D

---

*[drake/automotive/automotive_simulator.cc, line 302 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKbZ1JGzIo-qFmV06e:-KVKo8I-o0joDz045vh0:boyexbg) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/automotive_simulator.cc#L302)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Something like this would be less verbose (using vector's initializer_list):
> ```
> auto position_mux = builder_->template AddSystem<systems::Multiplexer<T>>(
>     std::vector<int>{kRpyJointNumPos, num_reg_pos_dofs});
> ```
> 
</details>
Done. Thanks.

---

*[drake/automotive/automotive_simulator.cc, line 309 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKbLSJP8rDwRNszPlT:-KVKoBP8mzUCwKztJvFu:b-7vcmtf) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/automotive_simulator.cc#L309)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> FYI The _GeometryWorld_'s appearance cannot come fast enough for me, to remove the need for ghastly logic like this stuff.  Wow!
</details>
Can't wait!!

---

*[drake/automotive/automotive_simulator.cc, line 319 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVK_lA56v_tMtLpeHyC:-KVKpXtQee-FfCCCtzqu:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/automotive_simulator.cc#L319)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Probably even the version on `master` previously deserved a helper method for this block, to improve readability of Start().  This new rewrite definitely does.
</details>
Done.

---

*[drake/automotive/automotive_simulator.cc, line 320 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVK_xZ__lmEvxDE4x81:-KVKogXZeoD8_aRk9b4b:b-b6i9qz) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/automotive_simulator.cc#L320)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Are we sure that calling `compile()` _after_ all of the bare-integer math above is sound?  If so, please clarify why in a comment.  If not, consider adding a TODO or assertion highlighting the hazard?
</details>
I moved this line to the beginning of this method and added a comment explaining why it's OK to do so.

---

*[drake/automotive/automotive_simulator.h, line 111 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKYCBsPsfHlNb8u0R-:-KVKtCSVco7elVRAgDUG:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/automotive_simulator.h#L111)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Don't use `const` on pass-by-value.
</details>
Done.

---

*[drake/automotive/automotive_simulator.h, line 120 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKYN3v177KD47oMqbJ:-KVKtMERjp8PkbznBAvP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/automotive_simulator.h#L120)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW The API comment reads like a `const` method, but its not marked const.
</details>
Done.

---

*[drake/automotive/automotive_simulator.h, line 130 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKYj1Cm5zKs7ODwPze:-KVKtgV1se2SCzingKFB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/automotive_simulator.h#L130)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Given this new blank line and the one following, consider if there's a way to restore the clarity that the "For building" comment applies to a group of several member fields, e.g. `// === For building ===` or something.
</details>
Done.

---

*[drake/automotive/test/automotive_simulator_test.cc, line 59 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKbz0-POvwwCGl1ECv:-KVKpmnZhbaU8z81BJMm:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/test/automotive_simulator_test.cc#L59)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW "17 are" typo, I think?
</details>
Done.

---

*[drake/automotive/test/automotive_simulator_test.cc, line 61 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKcIUyfwfuzlsw7qoO:-KVKq2dEe8DGefmPVPl9:b-u9omzz) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/test/automotive_simulator_test.cc#L61)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> FYI Once the API allows selecting the SDF / URDF file again, I think the unit test was superior with a simpler model under test in the base cases (and then add new case for a multibody model).  Consider a TODO for fixing that in the next iteration?
</details>
Done. The comment was added to the description of this unit test.

---

*[drake/automotive/test/automotive_simulator_test.cc, line 132 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKcCS-_cxraYQ3CISb:-KVKrR68iiM40hbhqZYZ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/test/automotive_simulator_test.cc#L132)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Rule-of-3 on missing name for magic literal "17". 
</details>
Done.

---

*[drake/automotive/test/automotive_simulator_test.cc, line 134 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVKc6T_IBTlYP2_krk6:-KVKt4bwAfQtT3euYijl:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/2468a11d83f4d7a6c5143e05890c318c06f2a3f7/drake/automotive/test/automotive_simulator_test.cc#L134)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Please amend this line so that it continues to test that the second car got added successfully, not the second body on the first car.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-:-KVKu68cYL7l150EoCLd:beojt2b)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/automotive_simulator.cc, line 223 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVLQSvR3HaS3RZmdHbt:-KVLQSvR3HaS3RZmdHbu:b3600vf) ([raw file](https://github.com/robotlocomotion/drake/blob/7a4724c2cd628c5fe85da4aff9fbe59f89d5768a/drake/automotive/automotive_simulator.cc#L223)):*
> ```cpp
> 
> template <typename T>
> void AutomotiveSimulator<T>::ConnectJointStateSourcesToVisualizer() {
> ```

BTW Consider `DRAKE_DEMAND(!started_);` here for additional clarity.

---

*[drake/automotive/test/automotive_simulator_test.cc, line 37 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVLQyAjtbY25Vw01SFq:-KVLQyAkMCOzAmjmImzO:bpnwdee) ([raw file](https://github.com/robotlocomotion/drake/blob/7a4724c2cd628c5fe85da4aff9fbe59f89d5768a/drake/automotive/test/automotive_simulator_test.cc#L37)):*
> ```cpp
> // takes as input a model and evaluates a SimpleCar based on the provided model.
> // This would allow, for example, models containing varying numbers of bodies to
> // be efficiently evaluated.
> ```

The first sentence says to me "Once we support different objects to render as the simple car, then test them here, but be sure not to duplicate lots of code."  Is that a correct interpretation?

The second sentence I can't make heads or tails of.  What is "efficiently" stated with respect to, I think is my main confusion?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-:-KVLRMN1zYEcHfKznqrd:bnfp4nl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@jwnimmer-tri I addressed your comments. PTAL. Thanks.

---

Review status: 3 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/automotive_simulator.cc, line 223 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVLQSvR3HaS3RZmdHbt:-KVLmyUltKRFSVLEA4ag:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7a4724c2cd628c5fe85da4aff9fbe59f89d5768a/drake/automotive/automotive_simulator.cc#L223)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Consider `DRAKE_DEMAND(!started_);` here for additional clarity.
</details>
Done.

---

*[drake/automotive/test/automotive_simulator_test.cc, line 37 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVLQyAjtbY25Vw01SFq:-KVLn5GH3Q4tCVmGz5KO:bfg2k8z) ([raw file](https://github.com/robotlocomotion/drake/blob/7a4724c2cd628c5fe85da4aff9fbe59f89d5768a/drake/automotive/test/automotive_simulator_test.cc#L37)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> The first sentence says to me "Once we support different objects to render as the simple car, then test them here, but be sure not to duplicate lots of code."  Is that a correct interpretation?
> 
> The second sentence I can't make heads or tails of.  What is "efficiently" stated with respect to, I think is my main confusion?
</details>
My intent was to use "efficiency" in terms of the number of lines of code.

I've updated this comment. PTAL. Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-:-KVLoIsYZNMVRZCyiWng:b-czr5yx)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3933)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3933)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
General question (for my own erudition, probably):  why AddSdfModel(), rather than AddUrdfModel()?  (Are we generally deprecating URDF in favor of SDF?  I recall some talk of that months and months ago, maybe, but URDF's are still hanging around....)

---

Reviewed 2 of 5 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/automotive_simulator.cc, line 117 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVW0anu_1-YlMcEAwMY:-KVW0anu_1-YlMcEAwMZ:birvmgy) ([raw file](https://github.com/robotlocomotion/drake/blob/df3457f05269922d2062b99ad8ab8d3233221980/drake/automotive/automotive_simulator.cc#L117)):*
> ```cpp
> 
>   const int model_instance_id = table.begin()->second;
>   const std::vector<const RigidBody*> bodies =
> ```

This variable `bodies` is no longer used.  (I'm surprised gcc didn't gripe about this.)

---

*[drake/automotive/automotive_simulator.h, line 115 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVVys_-8QXX4iFj-qbu:-KVVys_-8QXX4iFj-qbv:b-e2k4pz) ([raw file](https://github.com/robotlocomotion/drake/blob/df3457f05269922d2062b99ad8ab8d3233221980/drake/automotive/automotive_simulator.h#L115)):*
> ```objc
> 
>   // Connects the systems that output the pose of each vehicle to the
>   // visualizer. This is done by using multiplexers to add zero state sources
> ```

Unclear what "zero state sources" means.  And, in the following line, the comma breaks the clauses in such a way that makes it hard to understand that "zero state sources" are use for both uncontrolled position elements *and* velocity elements.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-:-KVW-8inNVXl5f0KSCdv:b7y5rl9)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Yes, URDF is no longer being developed, so I believe it can be considered deprecated. I believe our long-term goal is to switch to SDF.

I addressed your comments. PTAL.

---

Review status: 1 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/automotive_simulator.cc, line 117 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVW0anu_1-YlMcEAwMY:-KVWsrex55nd3fpl__2f:b3ykx85) ([raw file](https://github.com/robotlocomotion/drake/blob/df3457f05269922d2062b99ad8ab8d3233221980/drake/automotive/automotive_simulator.cc#L117)):*
<details><summary><i>Previously, maddog-tri wrote…</i></summary>
> This variable `bodies` is no longer used.  (I'm surprised gcc didn't gripe about this.)
</details>
Good catch! Done. Thanks.

---

*[drake/automotive/automotive_simulator.h, line 115 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-KVVys_-8QXX4iFj-qbu:-KVWskqlVoz1CgynzJiF:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/df3457f05269922d2062b99ad8ab8d3233221980/drake/automotive/automotive_simulator.h#L115)):*
<details><summary><i>Previously, maddog-tri wrote…</i></summary>
> Unclear what "zero state sources" means.  And, in the following line, the comma breaks the clauses in such a way that makes it hard to understand that "zero state sources" are use for both uncontrolled position elements *and* velocity elements.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-:-KVWsvX4WeGZoi39mJqx:bdvcjbe)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 4 of 4 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3933)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r6, 4 of 4 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3933#-:-KVXCLpLCzuH3nu6whx_:bnfp4nl)*
<!-- Sent from Reviewable.io -->
